Texture tiling

guiwitz opened the pull request:
# References and relevant issues

Aims to close https://github.com/napari/napari/issues/7149

# Description
This PR is an attempt at fixing at least partially the following issue: images that are not multiscale and that have one pixel dimension larger than the GL_MAX_TEXTURE_SIZE are simply downscaled to be rendered in napari. The consequence is that even when zooming, one can never see the full resolution of such an image.

For the 2D case, one solution to that problem is to break the images into tiles whose size is smaller than GL_MAX_TEXTURE_SIZE. As napari doesn't have a tiling mechanisms, this tiling needs to happen at the vispy nodes level. The solution proposed here is a new vispy node called `TiledImageNode` which is composed of a set of `Image` nodes. The `TiledImageNode` is used in a similar way as the `ImageNode` and `VolumeNode` i.e. the `ImageLayerNode` picks the correct type of node given the dimensionality and data size. This choice is made in the `get_node` function of `ImageLayerNode`. That function had to be modified to accept the data size as parameter so that the check of size could be done.

**Remaining issues**:
- The main issue is the conservation of the correct node upon updating the viewer. At the moment the `TiledImageNode` eventually gets replaced by the default `ImageLayerNode`. The `TiledImageNode` children are conserved but reattached to the `ImageLayerNode`. Therefore when changing e.g. the colormap, the change does not get propagated to children as the `ImageLayerNode` is not designed for that. The children can be explicitly set but this is not a good solution. The problem happens somewhere [here](https://github.com/guiwitz/napari/blob/ee64c142edc4f716155e3cbcff130fb3b168e155/src/napari/_vispy/layers/scalar_field.py#L73) because of a bad update sequence during which a node update happens while data are `None` (and hence their size below the tiled threshold)
- Currently the size check is done on the original data shape but it should be done on the sliced data

guiwitz committed:
tiled image implementation

guiwitz committed:
only create tiles for large images

guiwitz committed:
remove children when updating

guiwitz committed:
move transform to correct place

guiwitz committed:
move tile to ImageLayerNode

guiwitz committed:
add logic to switch image and tile

guiwitz committed:
remove comment

guiwitz committed:
avoid reset data, enforce layer properties

guiwitz committed:
Merge branch 'napari:main' into texture_tiling

guiwitz committed:
rename tiled image, print childs for checks, pass data to avoid update to Image

psobolewskiPhD commented:
I know it's WIP, but I wanted to stop in and say that This works! 🎉 
Pretty awesome, congrats!

But... 
I still have major reservations regarding memory/VRAM usage of this approach vs. auto-multiscale.
I'm on Apple Silicon, so unified memory, but OpenGL doesn't leverage that, so there is a copy involved.

Here's my test setup: set up a viewer and load a modest, but larger than GL limits image shape: (32914, 46000, 3) using tifffile.imread (just loading to numpy `data = imread...` no layer yet)
This makes Python interpreter memory go from 0.4Gb to 4.6GB
Then, with main:
- add_image, auto downsampled: memory jumps to 5.3GB, so we can assume ~0.7GB of VRAM? (it took ~1s)
    - Note: deleting the layer doesn't free the memory, need to close the viewer.
- add_image([data, data[::2, ::2, :], data[::4, ::4, :]]), make multiscale, memory jumps to: 5.1GB, so assume 0.5GB VRAM (it was basically instant)

With this PR:
- add_image, no downsample/multiscale: memory jumps to: 10.5GB, so we can assume ~6GB VRAM? (takes ~18 s FYI)

So that's my concern, based on having used my script that implements something like this. The VRAM usage can quickly become excessive I think.
It feels like it would need to be optional, to me at least, with auto-multiscale the default?

jni commented:
Just a shower thought, submitted here for @brisvag's consideration:

- the issue we are encountering is that the overlays are children of the main node.
- So when we change nodes because we change the data, we have to transfer all the children over.
- So our tiled node with image nodes as children causes a mistake
- So, proposed solution: overlays should be *siblings* of a main node that has the "data" node as one of its children.

guiwitz committed:
remove forced props updates

guiwitz committed:
avoid chaning tiles' parents

guiwitz committed:
actually remove re-parenting

guiwitz commented:
I managed to avoid the issue with reparenting the tiled nodes by just checking the type of the children and skipping the tiles. I now also removed the "forced" updates of the cmap and contrast to have a clean state. With these two changes, the implementation is functional **except** for the initialization. At the moment it is necessary to go through one step of data change, e.g. one round of transpose, for the data to become visible and the rendering changeable. I didn't manage to figure out what the `on_data_change` exactly triggers that is not triggered before.

guiwitz committed:
remove redundant loop

guiwitz committed:
remove data change check

guiwitz committed:
remove intial properties setting, remove prints

guiwitz committed:
avoid tile recreation

guiwitz committed:
simplify logic of tile use

guiwitz committed:
fix get_node singature for label layer

guiwitz committed:
check that node is tile instead of downscale

guiwitz committed:
fix typing

guiwitz committed:
add citation

guiwitz committed:
fix typing

guiwitz committed:
fix typing again

guiwitz committed:
no need to check data shape at init

guiwitz commented:
@jni I think I fixed most of the issues. Two tests are failing but my impression is that the failures are unrelated to the PR itself, but maybe you can have a look.

I modified [the test](https://github.com/napari/napari/pull/8395/commits/93033d6cb1e960836f3980d303b4bcf7f07ee8bd) that used to check if very large images were downscaled. I'm just checking that in that case the `node` is indeed a `TiledImageNode` but maybe we can do better.

I also wasn't sure how I should update my branch: update with rebase or with merge commit?

Finally, I see now that, even though I didn't touch the labels layer, I get an error message when trying to just display a large label image. I will fix that and hopefully then this PR will be out of WIP.

guiwitz committed:
add back downscaling for labels

guiwitz commented:
@jni I noticed I get the same behavior on main. I went a few versions back and it seems the problem has appeared at ~~napari=0.6.1~~ napari=0.6.2. So it seems unrelated to this PR.

jni commented:
Hmmm, I don't see the same behaviour on main... Are you sure??

Oh wait! I am on latest main, which includes #8423, which fixes the problem!!! Merging main into this branch also fixes it! Sorry for the noise! But that's great! 🤩

jni committed:
Merge branch 'main' into texture_tiling

jni commented on src/napari/_vispy/layers/scalar_field.py:
```diff
@@ -91,33 +101,45 @@ def _on_display_change(self, data=None) -> None:
         ):
             self._on_display_change(data)
         else:
+            # if not  isinstance(node, TiledImageNode):
```
```suggestion
```

guiwitz commented on src/napari/_vispy/layers/image.py:
```diff
@@ -43,19 +50,37 @@ def __init__(
+            else:
+                res = self._tiledimage_node
+        else:
+            res = self._volume_node
```
I very rarely use (if ever?) match statements, but I agree that here it makes the logic more readable!

guiwitz committed:
replace ifs with match

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
comment node selection function

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
clean up

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
comment TiledImageNode class

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
comments

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

guiwitz committed:
simplify logic

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
simplify tiling logic

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

guiwitz committed:
clean-up after comment integration

TimMonko commented on src/napari/_vispy/_tests/test_vispy_big_images.py:
```diff
@@ -12,8 +14,9 @@ def test_big_2D_image(make_napari_viewer):
     if visual.MAX_TEXTURE_SIZE_2D is not None:
-        s = np.ceil(np.divide(shape, visual.MAX_TEXTURE_SIZE_2D)).astype(int)
-        np.testing.assert_array_equal(layer._transforms['tile2data'].scale, s)
+        assert isinstance(visual.node, TiledImageNode)
```
I do think it's worth adding at least some kind of tests now that we are at the near final implementation. Maybe checking various attributes about the layer

TimMonko commented:
My bloody mouse isn't dragging right, so originally submitted the review. I'll take that opportunity to summarize some loose ends:

1. Grid mode is broken. Images seem to be transformed in space, but then escape the viewboxes, or something (I thought these were discrete and separate). Note how you can fully drag an image in the right grid visibly through the entire canvas. Colorbar is also broken in grid (but seem ok in stack mode), where they don't all appear. Another way that this was observed was in the single image script with then adding a labels layer (was downsampled, ofc) and then transforms were also not correct. @brisvag since you are the most recent to work on both grid and overlays, @guiwitz agreed that it would be great if you could take a look and potentially submit any fixes or work with him directly.
2. Quiet crashing when RAM is exceeded. It would be nice if we could somehow prevent such a bad outcome, at minimum a warning would be superb. 
3. Roll/transpose lead to a surprising result, but my brain doesn't wrap around this concept very well.

Otherwise, I think this is surprisingly smooth and performant, even with trying to paint a downsampled labels on top of a tiled image. I've run into this texture limit many times, especially on laptops without dedicated GPUs, but would have plenty of RAM to service this texture tiling nicely. I tested various layer attributes (i.e. opacity, blending, gamma, scale, etc) and all seem great, including with multiple layers. Toggling to other mods, like 3D

Follow-ups:
1. Add API to query a tiled property of the layer
2. Add documentation explaining a few things, like when this is triggered, how VRAM/max text size is different on different machines, and the limitations with respect to RAM

jni commented:
Ok I finally got a chance to play with the grid mode issue. It looks like for some reason the tiled image is not respecting the grid "limits", but normal images do. See this video:

https://github.com/user-attachments/assets/c581df6f-1a8a-47f8-ba39-fc3db4e8d9b3

Note: it was a complete coincidence that I picked "astronaut" to travel through the galaxy, but now I love it. 😂 

Note also how the astronaut correctly disappears when it reaches the middle of the screen, while the galaxy spills over across the whole screen.

@brisvag I presume this is because we're setting something incorrectly (or rather, passing through a getattr to the first child node instead of returning the "aggregated" answer) in the parent node... Any immediate thoughts?

brisvag commented:
Looking into the machinery behind, it looks like indeed there is some clipping based on bounds:

https://github.com/vispy/vispy/blob/25ef9915ae3affb499f2560fa6d73ebf0be20b06/vispy/scene/widgets/viewbox.py#L54-L57

We might indeed have an issue in vispy with reporting bounds correctly across the scenegraph. If it's not a python issue, then we need to go dig deeper into the shader implementation of this `Clipper` class...

jni committed:
Use Compound instead of Node

As proposed in:
https://github.com/napari/napari/pull/8395\#issuecomment-3776882524

Note Compound is a frozen class so we need to unfreeze in the
constructor and add any attributes that will be added/used later.

jni committed:
Pass custom kernel and interpolation to child nodes

jni committed:
Merge branch 'main' into texture_tiling

jni committed:
Fix typing

brisvag commented on src/napari/_vispy/layers/tiled_image.py:
```diff
@@ -0,0 +1,147 @@
+        self._subvisuals.append(visual)
+
+    def _transform_changed(self, event: Any = None) -> None:
+        BaseVisual._transform_changed(self)
```
We should add some comments about why thhese are needed

jni committed:
Add docstrings to add_subvisual and _transform_changed

... to explain why the methods are overridden.

jni committed:
Update test for texture tiling

jni commented:
I'm headed for bed.

There's still an issue with the axis transposition. Original image:

<img width="3178" height="2062" alt="screenshot_jJMIdgMd@2x" src="https://github.com/user-attachments/assets/5b2aaec4-5d63-42a6-acf8-c59ad0f72ab0" />

After hitting the transpose button:

<img width="3178" height="2062" alt="screenshot_4VCEx5Xm@2x" src="https://github.com/user-attachments/assets/30101daf-2db8-493d-b893-28061603087a" />

I think it should be a "simple" matter of finding the right event to hook up and flipping the tile offsets, but I couldn't quickly find the right place to do this. If anyone has a chance to look at it, it's appreciated. It could also be another thing to fix in a followup PR.

guiwitz committed:
fix transpose offsets

guiwitz committed:
connect tranpose to data update

guiwitz committed:
clean up

guiwitz commented:
@jni I tried to fix the issue with transpose. I think there's no layer event for such axis order changes, everything happens at the viewer level. So what I did was connecting the update related to transpose (and roll) to the `set_data` event of the layer. For the update itself, just resetting the `STTransform` seems sufficient, no need to change the offsets (not sure what vispy magic happens here). I'm not super confident this is the most elegant solution, and maybe I missed the actual proper event.

jni commented:
@TimMonko indeed we have only enabled this for 2D rendering. When you say "screwy" do you mean anything other than the downsample warning? (Which should be present on main also anyway.)
